Neos.ContentRepository.Core/Classes/ContentRepository.php

```diff
@@ -141,18 +143,24 @@ public function handle(CommandInterface $command): void
      */
     public function projectionState(string $projectionStateClassName): ProjectionStateInterface
     {
+        if (!isset($this->projectionStateCache)) {
+            foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
+                if ($projection instanceof ContentGraphProjectionInterface) {
+                    continue;
+                }
+                $projectionState = $projection->getState();
+                $this->projectionStateCache[$projectionState::class] = $projectionState;
+            }
+        }
         if (isset($this->projectionStateCache[$projectionStateClassName])) {
             /** @var T $projectionState */
             $projectionState = $this->projectionStateCache[$projectionStateClassName];
             return $projectionState;
         }
-        foreach ($this->projectionsAndCatchUpHooks->projections as $projection) {
-            $projectionState = $projection->getState();
-            if ($projectionState instanceof $projectionStateClassName) {
-                $this->projectionStateCache[$projectionStateClassName] = $projectionState;
-                return $projectionState;
-            }
+        if (in_array(ContentGraphReadModelInterface::class, class_implements($projectionStateClassName), true)) {
+            throw new \InvalidArgumentException(sprintf('Accessing the internal content repository projection state via %s(%s) is not allowed. Please use the API on the content repository instead.', __FUNCTION__, $projectionStateClassName), 1729338679);
```

Comment:
nitpick: why not move this check to the top and keep the rest like before?

Reply:
then it would be more expensive imo as we call `class_implements` always and here its just to improve the error message ;)

Reply:
as said above I would just allow it and thrash the check. it's a projectionstate after all, just one with _better_ API to reach it.

Reply:
We will have to disallow this with the [CR Privileges](https://github.com/neos/neos-development-collection/pull/5298) because they enforce the constraints on the CR instance

Reply:
right, ok fair enough. Then this place is fine for me.